compiler/src/dotty/tools/dotc/typer/Implicits.scala

```diff
@@ -1439,9 +1439,14 @@ trait Implicits:
              *  candidate that is strictly better than the failed candidate(s).
              *  If no such candidate is found, we propagate the ambiguity.
              */
-            def healAmbiguous(fail: SearchFailure, betterThanFailed: Candidate => Boolean) =
-              val newPending = remaining.filter(betterThanFailed)
-              rank(newPending, fail, Nil).recoverWith(_ => fail)
+            def healAmbiguous(fail: SearchFailure, ambiguousCands: List[RefAndLevel]) =
+              def betterThanAll(newCand: RefAndLevel, disambiguate: Boolean): Boolean =
+                ambiguousCands.forall(compareAlternatives(newCand, _, disambiguate) > 0)
+
+              val newPending = remaining.filter(betterThanAll(_, disambiguate = false))
```

Comment:
This looks like it does the right thing. But it does cost extra work, including in source versions where we don't even compare old and new implicits. Can we avoid that?

Reply:
Yes indeed, the check should be redundant if `!isPriorityChangeVersion`

I'll update that

Reply:
But even then we run the compareAlternatives twice...

Reply:
An alternative would to keep track of the nesting level in `SearchFailureType`. We would then be able to test between the possible origins of `healAmbiguous` in the case SearchFailure of `disambiguate`